### ISSUE TITLE
Excess semicolon lint produces wrong diagnostics with async main and derive macro

### ISSUE THREAD

**Issue by Kobzol** [rust-lang/rust#63947]
The new lint for excess semicolons (https://github.com/rust-lang/rust/pull/62984) has some weird behaviour when the lint is triggered inside an async function that is transformed using a procedural macro.

Rustc version: 1.39.0-nightly (9b91b9c10 2019-08-26)

Compiling this:
```rust
#[tokio::main]
async fn main() -> Result<(), ()> {
    let mut workers = 5;;
    Ok(())
}
```
gives this output:
```
warning: unused variable: `workers`
 --> examples/test.rs:1:1
  |
1 | #[tokio::main]
  | ^ help: consider prefixing with an underscore: `_workers`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: variable does not need to be mutable
 --> examples/test.rs:1:1
  |
1 | #[tokio::main]
  | ^ help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default
```
note how the spans are all targeting the procedural macro instead of the correct line.

If I remove the procedural macro or if I remove the excess semicolon, the spans are fine. This leads me to believe that it's an error in the lint, not in the macro, but maybe I'm wrong.

**Comment by Kobzol** [comment]
Without the excess semicolon the spans for the other warnings are correct even with the macro, which lead me to believe that the macro itself is not the problem. Does that mean that the macro somehow needs to support this new lint (or some family of lints)?

**Comment by Kobzol** [comment]
At my setup they trigger always (unless the compiler does not nothing and prints nothing out), however the combination `async` + macro + excess semicolon messes up the spans. I looked at the macro and tried to use `quote_spanned`, that didn't help.
